File: Class/Practise 5.1/src/MegaDataPool.cpp
#include "MegaDataPool.hpp"
#include <stdexcept>
#include <algorithm>

MegaDataPool::MegaDataPool(size_t poolSize)
    : poolSize(poolSize), pool(poolSize), used(poolSize, false) {
    for (size_t i = 0; i < poolSize; ++i) {
        pool[i] = std::make_shared<MegaData>();
    }
}

std::shared_ptr<MegaData> MegaDataPool::acquire() {
    std::lock_guard<std::mutex> lock(poolMutex);
    for (size_t i = 0; i < poolSize; ++i) {
        if (!used[i]) {
            used[i] = true;
            auto ptr = pool[i];
            return std::shared_ptr<MegaData>(ptr.get(), [this, i](MegaData* ptr) { release(i); });
        }
    }
    throw std::runtime_error("No available objects in the pool");
}

void MegaDataPool::release(size_t index) {
    std::lock_guard<std::mutex> lock(poolMutex);
    if (index >= poolSize) {
        throw std::out_of_range("Invalid index for release");
    }
    used[index] = false;
    pool[index]->reset();
}

size_t MegaDataPool::size() const {
    return poolSize;
}

size_t MegaDataPool::usedSize() const {
    std::lock_guard<std::mutex> lock(poolMutex);
    return std::count_if(used.begin(), used.end(), [](bool flag) { return flag; });
}

void MegaDataPool::resetAll() {
    std::lock_guard<std::mutex> lock(poolMutex);
    for (auto& data : pool) {
        data->reset();
    }
}

File: Home/Home 5.1/MegaDataPool.swift
import Foundation

enum MegaDataPoolError: Error {
    case noAvailableObjects
    case invalidObject
}

class MegaDataPool {
    private let poolSize: Int
    private var pool: [MegaData]
    private var used: [Bool]

    init(poolSize: Int) {
        self.poolSize = poolSize
        self.pool = Array(repeating: MegaData(), count: poolSize)
        self.used = Array(repeating: false, count: poolSize)
    }

    func acquire() throws -> MegaData {
        for i in 0..<poolSize {
            if !used[i] {
                used[i] = true
                return pool[i]
            }
        }
        throw MegaDataPoolError.noAvailableObjects
    }

    func release(_ data: MegaData) throws {
        if let index = pool.firstIndex(where: { $0 === data }) {
            used[index] = false
            pool[index].reset()
        } else {
            throw MegaDataPoolError.invalidObject
        }
    }

    func size() -> Int {
        return poolSize
    }

    func usedSize() -> Int {
        return used.filter { $0 }.count
    }
}
